ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFMkCollectionEvaluator.java

```diff
@@ -62,22 +62,16 @@ public ObjectInspector init(Mode m, ObjectInspector[] parameters)
     super.init(m, parameters);
     // init output object inspectors
     // The output of a partial aggregation is a list
-    if (m == Mode.PARTIAL1) {
+    if (mode == Mode.PARTIAL1 || mode == Mode.COMPLETE) {
       inputOI = parameters[0];
       return ObjectInspectorFactory.getStandardListObjectInspector(
           ObjectInspectorUtils.getStandardObjectInspector(inputOI));
     } else {
-      if (!(parameters[0] instanceof ListObjectInspector)) {
-        //no map aggregation.
-        inputOI = ObjectInspectorUtils.getStandardObjectInspector(parameters[0]);
-        return ObjectInspectorFactory.getStandardListObjectInspector(inputOI);
-      } else {
-        internalMergeOI = (ListObjectInspector) parameters[0];
-        inputOI = internalMergeOI.getListElementObjectInspector();
-        loi = (StandardListObjectInspector)
-            ObjectInspectorUtils.getStandardObjectInspector(internalMergeOI);
-        return loi;
-      }
+      internalMergeOI = (ListObjectInspector) parameters[0];
```

Comment:
is it safe to remove
`!(parameters[0] instanceof ListObjectInspector)`
condition?

Reply:
Yes, I think that condition was just to check "no map aggregation" or not, not for safe type casting.
As comments said above, "The output of a partial aggregation is a list"

https://github.com/apache/hive/blob/588c2ad9bd343b0ec1e6fa761bc36c0e00682ee6/ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFMkCollectionEvaluator.java#L43-L45

Reply:
I also guess so. `COLLECT_LIST|SET` processes objects of a type T like that.
- `Mode.PARTIAL1` or `Mode.COMPLETE`: T => List[T]
- `Mode.PARTIAL2` or `Mode.FINAL`: List[T] => List[T]

The original one tried to check the type of inputs, T or List[T], to conclude the current mode. But it can lead to a wrong conclusion when T is List[X] because both `List[X]` and `List[List[X]]` hit the condition.

Reply:
👍, thanks for the explanation!

Reply:
could we update the comment with what @okumin wrote:
````
Mode.PARTIAL1 or Mode.COMPLETE: T => List[T]
Mode.PARTIAL2 or Mode.FINAL: List[T] => List[T]
````
`The output of a partial aggregation is a list` is kinda confusing, PARTIAL2 is also partial

see comments at the top
// For PARTIAL1 and COMPLETE: ObjectInspectors for original data
// For PARTIAL2 and FINAL: ObjectInspectors for partial aggregations (list of objs)

Reply:
https://github.com/apache/hive/pull/5323/commits/74e7ce7087682c2a9c23a25b9c1423840baf3b0a

Comments updated. (I don't know this comments are necessary though.)

Reply:
@JeongDaeKim , sorry fo the confusion, no need for `// The output of a partial aggregation is a list, output of COMPLETE is List as well`. Just keep the mapping explanation